src/rrdp.rs

```diff
@@ -1,456 +1,962 @@
+            Err(anyhow!(format!(
+                "Cannot apply delta for serial '{}' to state serial '{}'",
+                delta_info.serial(),
+                self.serial()
+            )))
+        } else {
+            info!("Applying delta for serial: {}", delta_info.serial());
+            let delta = fetcher.read_delta_file(delta_info)?;
+            let delta_state = DeltaState::create(&delta);
+            self.current_objects.apply_delta(delta)?;
+            self.deltas.push_front(delta_state);
+            self.serial += 1;
+            Ok(())
+        }
+    }
+
+    /// Reads all deltas from a notification file, and returns a sorted
+    /// VecDeque of DeltaInfo. Used when the actual RRDP current state is
+    /// derived from a snapshot, but all deltas on notification file need
+    /// to be downloaded so that they may be made available.
+    fn read_deltas(deltas: &[DeltaInfo], fetcher: &Fetcher) -> Result<VecDeque<DeltaState>> {
+        let mut res = VecDeque::new();
+
+        for delta in deltas {
+            debug!("Read delta from {}", delta.uri());
+            let delta = fetcher.read_delta_file(delta)?;
+            let delta = DeltaState::create(&delta);
+            res.push_front(delta);
+        }
 
-impl From<XmlDecodeError> for SnapshotParserError {
-    fn from(err: XmlDecodeError) -> Self {
-        SnapshotParserError::Xml(err)
+        Ok(res)
     }
-}
 
-impl ProcessSnapshot for SnapshotParser {
-    type Err = SnapshotParserError;
+    /// Update state using a new snapshot
+    fn apply_snapshot(
+        &mut self,
+        notification_file: NotificationFile,
+        fetcher: &Fetcher,
+    ) -> Result<()> {
+        self.deprecate_snapshot_file();
+        self.deprecate_deltas_before(self.serial + 1);
+
+        self.serial = notification_file.serial();
+        self.session_id = notification_file.session_id(); // could have been reset.
+
+        self.current_objects =
+            CurrentObjectMap::read_snapshot(notification_file.snapshot(), fetcher)?;
+        self.snapshot = self
+            .current_objects
+            .derive_snapshot(self.session_id(), self.serial());
+        self.deltas = Self::read_deltas(notification_file.deltas(), fetcher)?;
 
-    fn meta(&mut self, session_id: Uuid, serial: u64) -> Result<(), Self::Err> {
-        self.meta = Some(ParsedMeta { session_id, serial });
         Ok(())
     }
 
-    fn publish(&mut self, uri: uri::Rsync, data: Vec<u8>) -> Result<(), Self::Err> {
-        self.publish.insert(ComparableRsyncUri(uri), data);
+    /// Write out all *missing* RRDP files. Optionally delay writing the notification file for
+    /// the specified number of seconds
+    pub fn write_rrdp_files(&self, notify_delay: u64) -> Result<()> {
+        self.write_snapshot()?;
+        self.write_missing_deltas()?;
+
+        if notify_delay > 0 {
+            info!(
+                "Waiting {} seconds before writing RRDP notification file",
```

Comment:
Should we say here waiting "at least {} seconds" (as the docs say sleep will return in no less than the specified seconds but may return after longer than requested?

Reply:
Unless you run this on a very anaemic system this should be close enough to the truth in my opinion.